best-first-cv/script.whizzml

```diff
@@ -0,0 +1,561 @@
+;; target number, or you run out of features.
+(define (select-features dataset-id
+                         nfeatures
+                         objective-id
+                         options
+                         k-folds
+                         pre-selected)
+  (let (obj-id (get-objective dataset-id objective-id)
+        k-fold-datasets (create-k-folds dataset-id k-folds)
+        input-ids (default-inputs dataset-id obj-id)
+        important-ids (important-fields dataset-id obj-id)
+        _ (log-info important-ids)
+        potentials (make-complementary important-ids pre-selected))
+    (loop (selected pre-selected
+           potentials potentials
+           queue [])
+
+      (if (or (>= (count selected) nfeatures) (empty? potentials))
+        (response dataset-id k-fold-datasets selected queue)
+        (let (_ (log-info "Making new candidates...")
+              candidates (make-candidates selected
+                                          (map (lambda (x) (x "feature"))
+                                          potentials))
+              _ (log-info "Selecting feature..." candidates)
+              next-feat-info (select-feature k-fold-datasets
+                                             obj-id
+                                             options
+                                             candidates
+                                             potentials)
+              _ (log-info "Next selected feature: " (next-feat-info "feature")))
+          (recur (append selected (next-feat-info "feature"))
+                 (make-complementary potentials [(next-feat-info "feature")])
+                 (cons {"features" (append selected
+                                           (next-feat-info "feature"))
+                              "phi" (next-feat-info "phi")} queue)))))))
+
+;; A simple function to get the max value in a list
+(define (get-max xs) (reduce (lambda (x y) (if (> x y) x y)) (head xs) xs))
+
+
+;; Filters the features in one list to be complementary with the ones in
+;; another list
+(define (make-complementary y-list x-list)
+  (iterate (acc y-list x x-list)
+    (filter (lambda (y) (not (= x (y "feature")))) acc)))
+
+
+;; Clean the created datasets and deliver a human-readable response
+(define (response dataset-id k-fold-datasets selected queue)
+  (map safe-delete k-fold-datasets)
+  {"selected-fields" (feature-names dataset-id selected)
+   "iterations-info" queue})
+
+
+;; Get feature names given ids
+(define (feature-names dataset-id ids)
+  (let (fields ((fetch dataset-id) "fields"))
+    (map (lambda (id) (fields [id "name"])) ids)))
+
+;; Get the default set of input fields for this dataset (all preferred
+;; fields minus the objective field).
+(define (default-inputs dataset-id obj-id)
+  (let (fields ((fetch dataset-id) "fields")
+        fids (keys fields))
+    (filter (lambda (k) (and (fields [k "preferred"] false) (not (= obj-id k))))
```

Comment:
if a field doesn't have a "preferred" value, the default is `true`, so i think that should be `(fields [k "preferred"] true)` no?

Reply:
yes. Good catch!